Topic 1 /Backend : Flask-Babel first setup

imenattatra opened the pull request:
## What type of PR is this? 

- [x] Feature


## Description

This is part of the [translation project ](https://github.com/metr-systems/backlog/issues/3476). The backend of Redash, as well as its frontend, includes an important part that we need to translate also. In this PR we start by settings up the tool that we need and also a first configuration, later (in next PRs) we will be extracting text and adding its translation. 

## Some technical details

#### used packages : 
We use[ flask-babel ](https://python-babel.github.io/flask-babel/) that was added to the project with the command `poetry add Flask-Babel Babel`. We add also `Babel` dependency because we need it to complement the first one... they both complement each other in a flask based project. 

We will use `Babel` to extract translation and handle translation files and we will handle `Flask-Babel` to dynamically handle the language and for runtime translation of code between `_("")`. 

In other words `Babel` in backend here is like `i18next` in frontend and `flask-babel` in backend is like `react-i18next ` in frontend when we compare packages in backend to packages in frontend. 

#### commands used : 
The first time : 

- Extracting strings: `pybabel extract -F redash/babel.cfg -o redash/locales/messages.pot .`
- Initialize translation file for german : `pybabel init -i redash/locales/messages.pot -d redash/translations -l de`
- Compile translation `pybabel compile -d redash/translations`

The other times:

- Extracting strings with same command 
- Update .po files with `pybabel update -i redash/locales/messages.pot -d redash/translations`
- Compile translation with same command 

#### the configuration : 
- We are trying to put everything related to backend translation inside redash/redash(the backend) folder, this is why you find the translation related files there.
- We are setting "de" as default language temporarily until we work on setting the current language through a flag feature


This so far is translation "Login to redash" only

<img width="300" alt="Screenshot 2024-12-04 at 17 29 45" src="https://github.com/user-attachments/assets/a38601f6-4764-4aac-af39-21341496c753">


## How is this tested?

There is still no working translation here, so i just made sure that pytest is working.
- [x] Unit tests (pytest)

## What is next?
- Extracting an adding translation for backend 
- Work on the language flag feature 
- Deploy priotrity UI
- Finish with the rest of the translation
- Deploy the full translation project

Redash spricht bald Deutsch !
![](https://i.giphy.com/media/v1.Y2lkPTc5MGI3NjExcDNucjYzdzdhYWR5ZGdwaGptNHp2cWFjM3Vpc21odnFlNG9qZ2RoYiZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/PUnMmyKKmYGsuNCZNQ/giphy.gif)

imenattatra committed:
add flask-babel dependency

imenattatra committed:
first setup for flask-babel

imenattatra committed:
initialize translation file for german language

imenattatra committed:
fix Flask-Babel setup in redash create_app

imenattatra committed:
update babel cfg file

imenattatra committed:
temporary update default language to be de

imenattatra committed:
update translation up to now

helenalebreton commented:
@imenattatra do I have a way to look into it on staging ?

imenattatra commented:
> @imenattatra do I have a way to look into it on staging ?

Not yet, also there is only one sentence translated. The thing is i am having two different epics (one for backend and one for frontend) so i plan to deploy backend to staging after i merge the frontend epic (with only frontend priority things in the beginning).
I am not sure actually about this strategy , what do you think @thiagogds ?

thiagogds commented on redash/app.py:
```diff
@@ -40,6 +46,10 @@ def create_app():
 
+    # Initialize the Flask-Babel extension
+    babel = Babel(app)
+    babel.locale_selector_func = get_locale
```
Why did you decide to use this way instead of like the documentation in the class initialization?


```suggestion
    babel = Babel(app, locale_selector=get_locale)
```

imenattatra replied:
yeah it was a mistake, i fixed it here in this commit https://github.com/metr-systems/redash/pull/19/commits/d486655258eb0d7c1795ce409df033c1891d3c84 , thanks

thiagogds reviewed (changes requested):
You shouldn't commit the .mo file to the project. The compilation of the messages should happen during the docker build. I think it makes sense to do that now in this PR because it's a base change for using translations.

thiagogds commented:
@imenattatra I think it's better to keep a single epic, it's easier to track the work and deploy to staging

imenattatra commented:
> @imenattatra I think it's better to keep a single epic, it's easier to track the work and deploy to staging

- I updated `epic/ui-translation` by merging it with latest `metr-main `. `epic/ui-translation` by now is having all priority frontend translation + is up-to-date with our main.
- I updated this branch here `flask-babel-setup` with all changes in `epic/ui-translation`, so now when we deploy backend changes to staging we have all changes to check there.

Thanks

imenattatra committed:
fix Babel instanciation

imenattatra committed:
remove .mo file from commited files

imenattatra committed:
add .mo file to .gitignore

imenattatra commented:
> @imenattatra do I have a way to look into it on staging ?

@helenalebreton It is deployed to staging 🥳

imenattatra committed:
add translation compiling step to docker building steps
